### PR TITLE
Add HostedPartition.Fenced to ReplicaManager

### PR DESCRIPTION
**Pull request by rondagostino**
As we catch up on the metadata log during startup, and before we have performed log recovery, we will assign partitions to a new "Fenced" state in ReplicaManager.  This allows us to "remember" what we processed as we were catching up.  Then later, after log recovery is complete, but before the broker becomes unfenced, we will iterate through all of the fenced partitions and become leader/follower as required.

Everything is implemented except the last parts -- getting notification that it is time to unfence the partitions and the implementation of the actual unfencing.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 1]
```diff
@@ -571,6 +806,17 @@ class ReplicaManager(val config: KafkaConfig,
...
+      case HostedPartition.None | HostedPartition.Offline => None
+    }
+  }
+
+  // An iterator over all online partitions. This is a weakly consistent iterator; a partition made offline or fenced
+  // after the iterator has been constructed could still be returned by this iterator.
+  private def onlinePartitionsIterator: Iterator[Partition] = {
+    allPartitions.values.iterator.flatMap {
+      case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(_) => None
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:819]: nit: seems reasonable to consolidate these three cases since this function is only trying to catch online partitions

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -280,6 +286,210 @@ class ReplicaManager(val config: KafkaConfig,
 
   private var logDirFailureHandler: LogDirFailureHandler = null
 
+  // initially true when using a Raft-based metadata quorum, and may flip-flop thereafter;
+  // always false when using ZooKeeper
+  @volatile private var fenced: Boolean = zkClient.isEmpty
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:291]: I am a little confused about the relationships between `fenced`, `fencedPartitionStates`, and `allPartitions`. I thought the advantage of having a `Fenced` state is that we would not need a separate collection because we can represent a fenced partition inside `allPartitions`. On the other hand, if fencing is really a global state as indicated by the flag here, then perhaps we need neither the `Fenced` state nor `fencedPartitionStates`. We could change `getOnlinePartitionOrException` to check for `Online(partition) if !fenced`.

Basically my concern is whether we can have clearly enforced invariants about the internal state here.

Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:291]: I think we can eliminate `fencedPartitionStates` and keep the same information inside the `HostedPartition.Fenced` instance -- it does seem to be a waste to have both.

Regarding the `fenced` variable, I was thinking about how we would deal with becoming fenced during normal operations, well after we had started up and recovered our logs.  In that situation, I did not intend this to represent a state such that every partition we lead or follow immediately transitions into the fenced state.  For example, while we are fenced I think it is conceivable that we could continue to successfully fetch from other brokers for partitions that we believe we follow and successfully respond to fetch requests for partitions that we believe we lead.  I intended this flag to indicate that any metadata changes related to partitions that we receive while in this state should not be relied upon as being the latest information and should be remembered for later (and potentially updated along the way) until we are no longer fenced and we know we have the right information.  Basically, like at startup, the state indicates when we need to catch up on the metadata log before acting on any partition metadata we might receive while we are catching up.

So, to summarize what I was thinking:

- At startup it is true that every partition is fenced until after we initially catch up and recover logs, then we start leading/following
- Later on, if we become fenced, we could potentially still lead/follow for a while, but any new partition metadata that shows up isn't relied upon as being correct -- such partitions transition to the fenced state and the metadata is remembered for when we catch up again -- then we start leading/following as appropriate.
- There is a case where if we are fenced for too long it is possible that we might have to abandon our metadata entirely and fetch a new snapshot.  This would happen if we are at metadata log offset X, but while fenced the offset X+1 became unavailable as an independently-fetchable offset and is only available as part of a snapshot.  In this case we have to do a "soft restart" of sorts, and we have some reconciliation to do on our internal state.  This reconciliation could be done in a number of ways, but one way would be to stop all leading/following, clear `allPartitions`, set the `fenced` flag, and only clear the `fenced` flag after we have rwsd the latest snapshot and caught up.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 3]
```diff
@@ -611,24 +872,56 @@ class ReplicaManager(val config: KafkaConfig,
...
+        // during a partition reassignment if a produce request from the client is sent to a broker after
+        // the local replica has been deleted.
+        Left(Errors.NOT_LEADER_OR_FOLLOWER)
+
+      case HostedPartition.None =>
+        Left(Errors.UNKNOWN_TOPIC_OR_PARTITION)
+    }
+  }
+
+  def localOnlineLogOrException(topicPartition: TopicPartition): Log = {
```
Comment (lbradstreet) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:899]: Could we please add a jmh benchmark for the frequently called ReplicaManager log lookup methods? Many of these are called very frequently in hot paths such as the fetch API code or replica fetcher, and we need to make sure we don't regress as the code is already overly heavy - I had an experimental patch to cut down on the use of Either and it dropped the CPU time and garbage by half. Unfortunately I can't locate the experiment or I would point you to the benchmark. I'm fine if we don't change the approach here as long as we have the benchmark and we aren't regressing.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 4]
```diff
@@ -142,6 +141,10 @@ case class FetchPartitionData(error: Errors = Errors.NONE,
                               preferredReadReplica: Option[Int],
                               isReassignmentFetch: Boolean)
 
+case class FencedPartitionMetadata(partition: Partition,
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: seems like we can flatten this into `Fenced`. Otherwise it is a case class consisting of only another case class.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 5]
```diff
@@ -142,6 +141,10 @@ case class FetchPartitionData(error: Errors = Errors.NONE,
...
+                                   metadataPartition: MetadataPartition,
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: could we just call this field `metadata`? I think the `Partition` suffix is implied from the context.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 6]
```diff
@@ -1581,6 +1844,10 @@ class ReplicaManager(val config: KafkaConfig,
       else {
         Set.empty[Partition]
       }
+      stateChangeLogger.info(s"Fencing ${partitionsToBeFenced.size} partition(s)")
+      if (partitionsToBeFenced.nonEmpty) {
+        makeFenced(imageBuilder, partitionsToBeFenced, metadataOffset)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: What happens to the `onLeadershipChange` callback if the partition is fenced? Because we may not be applying the change until later, I don't think we can send the callback through `handleMetadataRecords`. As an alternative, perhaps we need to provide a way to register a (say) `LeadershipListener` which gets notified about all leadership changes.

On a related note, we might need a notification to the `GroupCoordinator` when a partition becomes fenced so that it can stop attempting to write to the log. If we did a `LeadershipListener`, we could have three callbacks: `onBecomeLeader`, `onBecomeFollower`, and `onBecomeFenced`.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 7]
```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
...
 
   def nonOfflinePartition(topicPartition: TopicPartition): Option[Partition] = {
     getPartition(topicPartition) match {
       case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
       case HostedPartition.None | HostedPartition.Offline => None
     }
   }
 
+  // not sure if any of the 9 invocations of the above nonOfflinePartition() should be diverted here
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:734]: Right. I don't think it's safe to expose the `Partition` object while it is in a fenced state. For example, `GroupMetadataManager` calls `Partition.appendRecordsToLeader` on the partition returned from `nonOfflinePartition`. But that would not be safe until the log has been initialized.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 8]
```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
...
+  def onlinePartition(topicPartition: TopicPartition): Option[Partition] = {
+    getPartition(topicPartition) match {
+      case HostedPartition.Online(partition) => Some(partition)
+      case _ => None
+    }
+  }
+
+  def fencedPartition(topicPartition: TopicPartition): Option[Partition] = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: the only caller of this does not need access to the `Partition` object. Could we have an `isFenced(TopicPartition)` method instead?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 9]
```diff
@@ -280,6 +290,135 @@ class ReplicaManager(val config: KafkaConfig,
...
+  stateChangeLogger.info(s"Fenced=$fenced")
+
+  def fence(): Unit = {
+    replicaStateChangeLock synchronized {
+      fenced = true
+      stateChangeLogger.info(s"Fenced")
+    }
+  }
+
+  def unfence(): Unit = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: There seems to be a bit of duplication in here. I think I'm trying to understand why we don't use the `makeFollowers` and `makeLeaders` helpers.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 10]
```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
...
-      case Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Fenced(metadata) => includeFenced && metadata.partition.isAddingReplica(replicaId)
       case _ => false
     }
   }
 
   // Visible for testing
-  def createPartition(topicPartition: TopicPartition): Partition = {
+  def createPartition(topicPartition: TopicPartition, fenced: Boolean = false,
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: It might be better to have a separate `createFencedPartition` so that we can make the argument expectations clear.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 11]
```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
     Option(allPartitions.get(topicPartition)).getOrElse(HostedPartition.None)
   }
 
-  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int): Boolean = {
+  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int, includeFenced: Boolean): Boolean = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: As far as I can tell, we don't have any use cases which call for `includeFenced=true`. Did I miss any? Perhaps we should just document that this returns false for fenced partitions.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 12]
```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
...
+        throw new KafkaStorageException(s"Partition $topicPartition is in an offline log directory")
+
+      case Left(error) =>
+        throw error.exception(s"Error while fetching partition state for $topicPartition")
+
+      case Right(partition) => partition
+    }
+  }
+
+  def getNonOfflinePartitionOrException(topicPartition: TopicPartition): Partition = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Similar to other comment. Can we refuse to expose `Partition` while it is fenced?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 13]
```diff
@@ -1861,21 +2130,53 @@ class ReplicaManager(val config: KafkaConfig,
...
-        stateChangeLogger.trace(s"Completed batch ${metadataOffset} become-follower " +
+        stateChangeLogger.trace(s"Completed batch $metadataOffset become-follower " +
           s"transition for partition ${partition.topicPartition} with new leader " +
           s"${partitionStates(partition).leaderId}")
       }
 
     partitionsToMakeFollower
   }
 
+  def makeDeferred(builder: MetadataImageBuilder,
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:2148]: Wondering if we need to complete purgatory operations on fencing as well.

**Comment by hachikuji** [comment]
In the current patch, the "deferred/fenced" state is not propagated to the `Partition` object. Say we are the leader for a partition and it becomes fenced while loading from the latest snapshot. Because it is unaware of its fenced status, it will happily continue accepting read/write operations as the leader for the partitions. This is fine from a correctness perspective because the leaders won't be able to make progress, but it seems like a less than ideal implementation. For example, the leader will be able to continue sending `AlterIsr` requests to the controller.

An alternative we can consider is pushing the "deferred/fenced" state into `Partition`. Then the `Partition` itself can refuse read/write operations and will be smart enough to know that it should not keep shrinking the ISR. Possibly in the end, both `ReplicaManager` and `Partition` will need some awareness of this state.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 6 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: I deferred the invocations of the `onLeadershipChange` callbacks so they get invoked after we apply the deferred changes.

Here's a counterargument to the suggestion that we notify when partition changes get deferred.  We cannot rely on any received sate that we defer -- it may or may not be the correct state.  We will eventually apply the correct state, but it may or may not be the initial state that we defer for a partition.  I wonder if we want to act upon what could be intermediate state.

What we have may not be the ideal implementation, but I wonder if it at least correct n that the coordinators would not be able to incorrectly append to an offset/txn state partition.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 7 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: > would not be safe until the log has been initialized

What is unsafe about it?  Looking at the code it throws `NotLeaderOrFollowerException` if the log doesn't exist, and if it does I assume it would be okay to append under the assumption that anything done incorrectly would not be committed.  Is this incorrect?

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 12 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Are we going to allow follower fetching when the broker is fenced? if so, then I wonder if it becomes necessary to expose the Partition even when we have deferred changes.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 14]
```diff
@@ -611,24 +885,56 @@ class ReplicaManager(val config: KafkaConfig,
...
-    nonOfflinePartition(topicPartition).flatMap(_.log)
+  // Indicates if the partition is online and has a future Log.
+  // Equivalent to futureLocalDeferredOrOnlineLogExists() when using ZooKeeper.
+  def futureLocalOnlineLogExists(topicPartition: TopicPartition): Boolean = {
+    deferredOrOnlinePartitionOrException(topicPartition).futureLog.isDefined
+  }
+
+  // Indicates if the partition is either online or deferred and has a future Log.
+  // Equivalent to futureLocalOnlineLogExists() when using ZooKeeper.
+  def futureLocalDeferredOrOnlineLogExists(topicPartition: TopicPartition): Boolean = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:920]: A few of these are unused. Can we cut this down to the minimal apis we need?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 15]
```diff
@@ -611,24 +885,56 @@ class ReplicaManager(val config: KafkaConfig,
...
-    getPartitionOrException(topicPartition).localLogOrException
+  // Returns the Log for the partition if it is online and has one, otherwise raises an exception.
+  // Equivalent to localDeferredOrOnlineLogOrException() when using ZooKeeper.
+  def localOnlineLogOrException(topicPartition: TopicPartition): Log = {
+    onlinePartitionOrException(topicPartition).localLogOrException
+  }
+
+  // Returns the Log for the partition if it is deferred or online and has one, otherwise raises an exception.
+  // Equivalent to localOnlineLogOrException() when using ZooKeeper.
+  def localDeferredOrOnlineLogOrException(topicPartition: TopicPartition): Log = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:896]: I didn't see a response to my previous comment, so I will repeat it here in case it was missed. I think we need to be clear about the contract when the partition is in the "deferred" state. It's unclear to me why we need to expose either the `Log` or `Partition` object at all in this state.  Keep in mind that in this patch when the state is "deferred," the leader/epoch information inside `Partition` is guaranteed to be out of date. It is only inside `applyDeferredMetadataChanges` that we apply the latest metadata to `Partition`. It may be safe to rely on the old state, but what is the point?

If ultimately we think this is the right implementation--that is if ultimately we just want to continue with the old state until the initialization/reloading has completed--then I'm not sure I see the benefit of mixing the deferred state into `allPartitions`. It would be simpler to collect the deferred entries in a separate map which would be applied at the end when the loading is complete. Then we could dump all of these "deferredOrOnline" APIs and potentially the duplication inside `applyDeferredMetadataChanges`.

Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:854]: Metadata gets deferred when the broker is fenced and new metadata arrives.  If new metadata is arriving then that probably means we are heartbeating successfully, and that likely means we'll be caught up soon.

I doubt we would end up in some intermediate state for very long.  I mean, it's possible, but it's unlikely -- we would have to be fenced, then back in communication, then out of communication.  The most likely event by far is that we have intermediate states for a short period of time.

So maybe the best thing to do is to assume the worst as soon as anything defers -- i.e. we don't expose the partition when it has deferred metadata.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 6 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: I've come around to assuming all bets are off once we are deferring partition state -- its easier to reason about, and the likeliest possibility by far is that we will xoon catch up and enter a consistent state.  This means ideally we should notify right away when a partition first gets deferred state, and the receiver of the callback can decide what to do.  Right now we only have `onLeadershipChange` with an indication if it's a new leader or a new follower.  We might need something new like you suggest.  However, since we are likely to be in a consistent state soon, maybe this is an optimization that can wait -- the "ideal" solution might not be to costly to defer (pun intended).

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 7 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: We now scope all "deferred" methods as private within ReplicaManager.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 12 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: If the broker is fenced then we won't be getting any new metadata, and nothing will defer.  Everything related to deferred partitions is now scoped private within ReplicaManager.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 16]
```diff
@@ -1779,101 +2062,167 @@ class ReplicaManager(val config: KafkaConfig,
...
   }
 
-  private def makeFollowers(builder: MetadataImageBuilder,
+  private def makeFollowers(prevPartitionsAlreadyExisting: Set[MetadataPartition],
+                            currentBrokers: MetadataBrokers,
                             partitionStates: Map[Partition, MetadataPartition],
                             highWatermarkCheckpoints: OffsetCheckpoints,
-                            metadataOffset: Long): Set[Partition] = {
+                            defaultMetadataOffset: Long,
+                            metadataOffsets: Map[Partition, Long] = Map.empty): Set[Partition] = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:2070]: It looks like we are only using this collection as well as `defaultMetadataOffset` for the purpose of logging. Is that right? We have one log message already in `makeDeferred` which indicates the offset when the update was received, so that makes me feel that this is probably overkill. Perhaps rather than indicating the exact offset in which the change was present, we can treat this more as the latest state at the time that the last offset was applied. Otherwise, having two collections keyed by `Partition` is suboptimal.

An alternative is that we could try to keep track of the offset in `MetadataPartition`.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 17]
```diff
@@ -1531,56 +1787,75 @@ class ReplicaManager(val config: KafkaConfig,
     val builder = imageBuilder.partitionsBuilder()
     val startMs = time.milliseconds()
     replicaStateChangeLock synchronized {
+      val deferrable = changesDeferrable
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:1790]: nit: seems unnecessary to cache this since we are in the lock anyway

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 18]
```diff
@@ -1531,56 +1787,75 @@ class ReplicaManager(val config: KafkaConfig,
...
+          stateChangeLogger.trace(s"Metadata batch $metadataOffset: locally changed: ${state}")
         }
         builder.localRemoved().foreach { state =>
-          stateChangeLogger.trace(s"Metadata batch ${metadataOffset}: locally removed: ${state}")
+          stateChangeLogger.trace(s"Metadata batch $metadataOffset: locally removed: ${state}")
         }
       }
       // First create the partition if it doesn't exist already
+      // partitionChangesToBeDeferred maps each partition to be deferred to its (current state, previous deferred state if any)
+      val partitionChangesToBeDeferred = mutable.HashMap[Partition, (MetadataPartition, Option[HostedPartition.Deferred])]()
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:1804]: The way we do this is a little curious. If `changesDeferrable` is true, then aren't `partitionsToBeLeader` and `partitionsToBeFollower` unnecessary? It seems like we can simplify this and just jump directly to `makeDeferred` in that case. That would make this code easier to follow since the cases would not all be mixed together.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 19]
```diff
@@ -2062,10 +2411,10 @@ class ReplicaManager(val config: KafkaConfig,
   }
 
   private def leaderPartitionsIterator: Iterator[Partition] =
-    nonOfflinePartitionsIterator.filter(_.leaderLogIfLocal.isDefined)
+    onlinePartitionsIterator.filter(_.leaderLogIfLocal.isDefined)
 
   def getLogEndOffset(topicPartition: TopicPartition): Option[Long] =
-    nonOfflinePartition(topicPartition).flatMap(_.leaderLogIfLocal.map(_.logEndOffset))
+    deferredOrOnlinePartition(topicPartition).flatMap(_.leaderLogIfLocal.map(_.logEndOffset))
```
[the remark this answers is not in the document]
Comment (rondagostino) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:2417]: I think this is resolved in the AK PR, but we will have to review and test carefully there.  https://github.com/apache/kafka/pull/10069

**Review by rondagostino — COMMENTED** [review]
This is superseded by the AK PR https://github.com/apache/kafka/pull/10069.

**Comment by rondagostino** [comment]
Closing in favor of the AK PR https://github.com/apache/kafka/pull/10069
